components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -67,71 +71,582 @@ fn sqlcipher_3_compat(conn: &Connection) -> Result<()> {
     Ok(())
 }
 
-fn sqlcipher_export(conn: &mut Connection, new_db_path: impl AsRef<Path>) -> Result<()> {
-    // Export sqlite data to a plaintext DB using the strategy from example 2 here:
-    // https://www.zetetic.net/sqlcipher/sqlcipher-api/index.html#sqlcipher_export
-    let path = new_db_path.as_ref().as_os_str();
-    let path = path
-        .to_str()
-        .ok_or_else(|| ErrorKind::InvalidPath(path.to_os_string()))?;
-
-    conn.execute("ATTACH DATABASE ? AS plaintext key ''", &[path])?;
-    // this one is a bit weird because it's a SELECT statement that we know will return 0 rows.
-    // rusqlite will return an Error if we use execute(), so we use query_row with a dummy closure
-    conn.query_row("SELECT sqlcipher_export('plaintext')", NO_PARAMS, |_| Ok(0))?;
-    conn.execute("DETACH DATABASE plaintext", NO_PARAMS)?;
+//Manually copy over row by row from sqlcipher db to a plaintext db
+pub fn migrate_from_sqlcipher_db(
+    cipher_conn: &mut Connection,
+    new_db_store: LoginStore,
+    encryption_key: &str,
+) -> Result<MigrationMetrics> {
+    // encrypt the username/password data
+    let encryptor = EncryptorDecryptor::new(encryption_key)?;
+
+    // Migrate tables separately due to specific columns in each needing
+    // to be ported over
+    let local_metrics = migrate_local_logins(&cipher_conn, &new_db_store, &encryptor)?;
```

Comment:
I'm wondering if the `migrate_mirror_logins` and `migrate_local_logins` functions should share the same transaction. I don't think we'd want either table to be migrated if the other isn't. IIRC in logins a combination of the records in the local and mirror table represent what was last synced so having them be in incongruous states seems undesirable. If they share this transaction, I would argue that `migrate_sync_metadata` should be included and metrics (with the exception of fixup metrics) should be collected for it.

Reply:
Hmmm that does raise a few more questions too, we also probably need to revisit the discussion of end-to-end flow of rollbacks (rollforwards?) and everything inbetween:

1. I initially had it as separate tx just for ease of getting this code out, but definitely agree that maybe sharing the same transaction may help with the potential rollback logic.
  - However, it will be rare for an entire DB to be corrupted and most likely it'll be a few rows that will not be able to be fixed up/migrating at that point, do we also skip if we find that guid in the other table? From my understanding, this will only happen if they have a sync account but is migrating before syncing.
  - Since we're doing row by row instead of the entire table, if we cannot fixup a row, we just skip it -- lessening the reliance on needing to do any kind of major rollback (unless something goes horribly wrong with trying to fixup that row)
2. Since this is kind of a one-way street kind of migration. We most likely need to have a discussion of how we want to proceed if we can't even get the DB to open (I think our current stance is, trash it and start fresh) but as mentioned above, since we are doing row-by-row there is less chance of really bad things happening since we can just skip that row and move on.

Reply:
> rollbacks (rollforwards?) 

LOL, right. Rolling back probably isn't a viable option. We'd have to migrate the database away from SQLCipher at some point and what would change between the time the rollback occurred and the next migration attempt that would increase the likelihood of success? So, yes, rollforwards I think.

>     1. I initially had it as separate tx just for ease of getting this code out, but definitely agree that maybe sharing the same transaction may help with the potential rollback logic.

Yeah after I suggested the shared transaction approach, I started second-guessing it. If we're not rolling back, and I don't think we should, then we'd either be skipping records that can't be ported over or trashing databases in the corrupted db scenario. 

However I think we do need to ensure that the tables are in sync (🤡) with each other. If that's the case, we could remove records in the corresponding tables if they can't be migrated as you suggested. However what's to stop us from resetting sync and forcing a sync after the migration? That's a Friday EOD question but I'm not deleting it.

Reply:
Yeah, I agree entirely with this - the problem doing local then the mirror is exactly that the tables might get out of sync. But the more I think about it, that *might* actually be an easily recoverable outcome: Consider:

* failing to migrate the `local` record but successfully migrating the `mirror` record: That means "local changes" are lost. Non sync users have lost this record entirely, Sync users will still work based on the mirror, but will have lost a local change waiting to sync up. That should be rare. I think the only problem here is that `is_overridden` would be wrong - so we should "migrate" that based on whether the actual record exists rather than the old DB value.

* successful `local` record but failing mirror record: should be impossible for non-sync users (they have no mirror records), and I think this would still work fine for sync users if we set `sync_status` to `SyncStatus::New/Changed`

A final consideration is: what happens when we've done a fixup? Does a fixup mean a "local change" (ie, that we should mark the item as locally modified and sync it up?) Fixing up a mirror record for a sync user almost certainly means a corrupt DB - that should exactly reflect the server - but we probably should not discard it entirely unless we have a 'local' record (see above, a "fixed up" local mirror record is better than discarding it because sync users, but we almost certainly want to arrange for us to re-download the mirror for that record)

I think the only way to capture all this (except metadata) as a single loop - ideally with a union statement, but if 2 read loops made it much easier that'd be ok. Then a struct/enum/handwave that represents the entire set of local states a GUID can be in and some rust pattern matching to deal with it. As as @lougeniaC64 mentions, a reasonable decision in some cases might be "also decline to migrate metadata, so we do a full server reconcile")